### PR TITLE
fix `The LiveNodeListWrapper imitates wrongly a native behavior` (close #1376)

### LINKED ISSUE #1376: The LiveNodeListWrapper imitates wrongly a native behavior
Example:
```js
const script = document.getElementsByTagName('script');

console.log(scripts.length);
// origin
// > 5
// proxy
// > 5

console.log(scripts.hasOwnProperty(0));
// origin
// > true
// proxy
// > false

const count = 0;

for (const prop in scripts)
    count++;
console.log(count);
// origin
// > 10
// proxy
// > 0
```
page https://www.usps.com/

### PR DESCRIPTION
**Pull request by LavrovArtem**

### CODE REVIEW

File: src/client/sandbox/node/live-node-list/html-collection-wrapper.js [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        configurable: true,
+        enumerable:   true,
+        get:          function () {
+            this.refreshHtmlCollection();
+
+            return this.filteredHtmlCollection.length;
+        }
+    },
+
+    refreshHtmlCollection: {
```
Comment (miherlosev) [inline src/client/sandbox/node/live-node-list/html-collection-wrapper.js]: It's an internal function. Rename to `_refreshHtmlCollection`

File: src/client/sandbox/node/live-node-list/html-collection-wrapper.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+                if (nameAttr !== null)
+                    currentNamedProps.push(nameAttr);
+            }
+
+            this.version = DOMMutationTracker.getVersion(this.tagName);
+
+            updateHtmlCollectionElementGetters(this, storedFilteredCollectionLength, this.filteredHtmlCollection.length);
+
+            if (currentNamedProps)
```
Comment (miherlosev) [inline src/client/sandbox/node/live-node-list/html-collection-wrapper.js]: Move this condition to the `updateNamedProps` function.

File: src/client/sandbox/node/live-node-list/html-collection-wrapper.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    refreshHtmlCollection: {
+        value: function () {
+            if (!DOMMutationTracker.isOutdated(this.tagName, this.version))
+                return;
+
+            const storedFilteredCollectionLength = this.filteredHtmlCollection.length;
+            const collectionLength               = nativeMethods.htmlCollectionLengthGetter.call(this.htmlCollection);
+            const currentNamedProps              = this.namedProps ? [] : null;
+
+            this.filteredHtmlCollection.length = 0;
```
Comment (miherlosev) [inline src/client/sandbox/node/live-node-list/html-collection-wrapper.js]: Extract this logic to the separate method:

```js
const { filteredCollection, namedProperties } = this._filterCollection();

```

File: src/client/sandbox/node/live-node-list/html-collection-wrapper.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+const ELEMENTS_WITH_NAME_ATTRIBUTE = ['button', 'fieldset', 'form', 'iframe',
+    'input', 'map', 'meta', 'object', 'output', 'param', 'select', 'textarea'];
+const DEFAULT_SHADOW_GETTERS_SHIFT = 10;
+let shadowGettersCount             = 0;
+
+export default function HTMLCollectionWrapper (htmlCollection, tagName) {
+    tagName = tagName.toLowerCase();
+
+    nativeMethods.objectDefineProperties.call(Object, this, {
+        htmlCollection:         { value: htmlCollection },
```
Comment (miherlosev) [inline src/client/sandbox/node/live-node-list/html-collection-wrapper.js]: `html` suffix is an unneeded context. We is already in `HTMLCollectionWrapper` context.
So, remove `html` suffix from variable and function names (`htmlCollection`. `refreshHmtlCollection` and etc.)

File: src/client/sandbox/node/live-node-list/html-collection-wrapper.js [hunk 5]
```diff
@@ -118,27 +98,58 @@ function updateHtmlCollectionElementGetters (wrapper, oldLength, currentLength)
     while (oldLength > currentLength)
         delete wrapper[--oldLength];
 
-    if (currentLength > shadowGettersCount - DEFAULT_SHADOW_GETTERS_SHIFT)
-        addShadowGetters(currentLength - (shadowGettersCount - DEFAULT_SHADOW_GETTERS_SHIFT));
+    if (currentLength > collectionProtoGettersCount - COLLECTION_PROTO_GETTERS_RESERVE)
```
Comment (miherlosev) [inline src/client/sandbox/node/live-node-list/html-collection-wrapper.js]: Extract `collectionProtoGettersCount - COLLECTION_PROTO_GETTERS_RESERVE` to variable.
I propose to call it as `prevCollectionProtoGettersCount`.

Comment (LavrovArtem) [reply src/client/sandbox/node/live-node-list/html-collection-wrapper.js]: What about `maxCollectionLength`?

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

File: src/client/sandbox/node/live-node-list/html-collection-wrapper.js [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    this._refreshCollection();
+
+    return this._filteredCollection[index];
+};
+
+if (HTMLCollection.prototype.namedItem) {
+    HTMLCollectionWrapper.prototype.namedItem = function (...args) {
+        this._refreshCollection();
+
+        const findNamedItem = this._collection.namedItem.apply(this._collection, args);
```
Comment (churkin) [inline src/client/sandbox/node/live-node-list/html-collection-wrapper.js]: `foundNamedItem`?

Comment (LavrovArtem) [reply src/client/sandbox/node/live-node-list/html-collection-wrapper.js]: I think just `namedItem`

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

**Pull request merged**